Binding for the communication with the KM200

Markinus opened the pull request:
Bindung for the communication with the KM200 Buderus Gateway Logamatic web KM200 ( KM100 / KM50 should work too, tested on KM200). It is possible to get and set string and float values. There are string, float, date and switch items supported.
https://www.buderus.de/de/produkte/catalogue/alle-produkte/7719_gateway-logamatic-web-km200-km100-km50

9037568 commented on bundles/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200Comm.java:
```diff
@@ -0,0 +1,712 @@
+                    val = provider.getParameter(item).get("on");
+                }
+            } else {
+                logger.error("witch-Item only on configured on/off string values:" + command.toString());
```
"Switch-Item"

Also, use parameterized logging instead of string concatenation

9037568 commented on bundles/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200Comm.java:
```diff
@@ -0,0 +1,712 @@
+            if (type.equals("stringValue")) {
+                nodeRoot = new JSONObject().put("value", val);
+            } else {
+                logger.error("Not supported type for SwitchItem:" + type.toString());
```
Use parameterized logging instead of string concatenation

9037568 commented on bundles/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200Comm.java:
```diff
@@ -0,0 +1,712 @@
+            }
+
+        } else {
+            logger.error("Bindingtype not supported : " + itemType.getClass().toString());
```
Use parameterized logging instead of string concatenation

9037568 commented on bundles/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200Comm.java:
```diff
@@ -0,0 +1,712 @@
+            logger.error("Couldn't find the right configuration");
+            return;
+        }
+        logger.debug("Encoding" + nodeRoot.toString());
```
Use parameterized logging instead of string concatenation

9037568 commented on bundles/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200CommObject.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    public void setValue(String val, String type) {
+        if (type == "stringValue") {
```
if(type.equals("stringValue"))

9037568 commented on bundles/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200CommObject.java:
```diff
@@ -0,0 +1,99 @@
+        if (type == "stringValue") {
+            value = new String(val);
+        }
+        if (type == "floatValue") {
```
if(type.equals("floatValue"))

9037568 commented on bundles/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200CommObject.java:
```diff
@@ -0,0 +1,99 @@
+            value = new String(val);
+        }
+        if (type == "floatValue") {
+            value = new Float(Float.valueOf(val));
```
the "new Float" should be unnecessary

9037568 commented on bundles/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200CommObject.java:
```diff
@@ -0,0 +1,99 @@
+        valueParameter = val;
+    }
+
+    public void setValue(String val, String type) {
```
Shouldn't this method have a try/catch for NumberFormatExceptions?

9037568 commented on bundles/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200CommObject.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    public Object getValue() {
+        if (value == null) {
```
This should be simplified to just:
>return value;

9037568 commented on bundles/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200CommObject.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    public Object getValueParameter() {
+        if (valueParameter == null) {
```
This should be simplified to just:
>return valueParameter;

9037568 commented on bundles/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200Device.java:
```diff
@@ -0,0 +1,257 @@
+
+class KM200Device {
+
+    private static final Logger logger = LoggerFactory.getLogger(KM200Binding.class);
```
This should be ... .getLogger(KM200Device.class)

9037568 commented on bundles/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200Device.java:
```diff
@@ -0,0 +1,257 @@
+            try {
+                md = MessageDigest.getInstance("MD5");
+            } catch (NoSuchAlgorithmException e) {
+                // TODO Auto-generated catch block
```
There should be a logger statement here

9037568 commented on bundles/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200Device.java:
```diff
@@ -0,0 +1,257 @@
+            try {
+                bytesOfGatewayPassword = gatewayPassword.getBytes("UTF-8");
+            } catch (UnsupportedEncodingException e) {
+                // TODO Auto-generated catch block
```
There should be a logger statement here

9037568 commented on bundles/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200Device.java:
```diff
@@ -0,0 +1,257 @@
+            try {
+                bytesOfPrivatePassword = privatePassword.getBytes("UTF-8");
+            } catch (UnsupportedEncodingException e) {
+                // TODO Auto-generated catch block
```
There should be a logger statement here

9037568 commented on bundles/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200Device.java:
```diff
@@ -0,0 +1,257 @@
+            for (KM200CommObject object : serviceMap.values()) {
+                if (object != null) {
+                    String val = "", type, valPara = "";
+                    logger.debug("List:" + object.getServiceType() + " " + object.getFullServiceName());
```
Use parameterized logging instead of string concatenation.

9037568 commented on bundles/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200Device.java:
```diff
@@ -0,0 +1,257 @@
+                                    valPara += ";";
+                                    valPara += valParas.get(1);
+                                } else {
+                                    logger.error("Value parameter for float != 2");
```
This should probably have a lower severity.

9037568 commented on bundles/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200Device.java:
```diff
@@ -0,0 +1,257 @@
+                        val = "";
+                        valPara = ";";
+                    }
+                    logger.info(object.getReadable().toString() + ";" + object.getWriteable().toString() + ";"
```
Use parameterized logging instead of string concatenation.

9037568 commented on bundles/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200Device.java:
```diff
@@ -0,0 +1,257 @@
+     *
+     * @since 1.9.0
+     */
+    public void listAllServices() {
```
How often is this method called? If it is (or can be) called frequently, these logger statements should probably be debug instead of info.

9037568 commented on bundles/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200GenericBindingProvider.java:
```diff
@@ -0,0 +1,212 @@
+        String service = null;
+        HashMap<String, String> parameterMap = new HashMap<String, String>();
+        /* Check whether some defined services are used */
+        logger.info("Bind Config: " + bindingConfig);
```
Use parameterized logging instead of string concatenation.

9037568 commented on bundles/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200GenericBindingProvider.java:
```diff
@@ -0,0 +1,212 @@
+        /* Available options for a SwitchItem are on:xxx and off:yyy with a allowed value xxx|yyy */
+
+        String[] keyTypeStructure = bindingConfig.split("\\s+");
+        logger.info("Bind Config nbr: " + keyTypeStructure.length);
```
Use parameterized logging instead of string concatenation.

9037568 commented on bundles/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200GenericBindingProvider.java:
```diff
@@ -0,0 +1,212 @@
+        String[] keyTypeStructure = bindingConfig.split("\\s+");
+        logger.info("Bind Config nbr: " + keyTypeStructure.length);
+        if (keyTypeStructure.length > 3) {
+            logger.error("Incorrect numer of structures in configuration string '" + bindingConfig + "'");
```
"number"

Also, use parameterized logging instead of string concatenation.

9037568 commented on bundles/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200GenericBindingProvider.java:
```diff
@@ -0,0 +1,212 @@
+        if (keyTypeStructure.length > 3) {
+            logger.error("Incorrect numer of structures in configuration string '" + bindingConfig + "'");
+            throw new BindingConfigParseException(
+                    "Incorrect numer of structures in configuration string '" + bindingConfig + "'");
```
"number"

9037568 commented on bundles/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200GenericBindingProvider.java:
```diff
@@ -0,0 +1,212 @@
+            String[] keyValueStructure = structure.split(":");
+
+            if (keyValueStructure.length != 2) {
+                logger.error("Incorrect key:value structure in configuration string '" + bindingConfig + "'");
```
Use parameterized logging instead of string concatenation.

9037568 commented on bundles/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200GenericBindingProvider.java:
```diff
@@ -0,0 +1,212 @@
+            if (key.equals("service")) {
+                service = value;
+                if (!service.contains("/")) {
+                    logger.error("Wrong service string without / in configuration string '" + bindingConfig + "'");
```
Use parameterized logging instead of string concatenation.

9037568 commented on bundles/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200GenericBindingProvider.java:
```diff
@@ -0,0 +1,212 @@
+            } else if (key.equals("on") || key.equals("off")) {
+                parameterMap.put(key, value);
+            } else {
+                logger.error("Unsupported key (" + key + ") in configuration string '" + bindingConfig + "'");
```
Use parameterized logging instead of string concatenation.

9037568 commented on bundles/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200GenericBindingProvider.java:
```diff
@@ -0,0 +1,212 @@
+        if (parameterMap.containsKey("on") || parameterMap.containsKey("off")) {
+            if (!(parameterMap.containsKey("on") && parameterMap.containsKey("off"))) {
+                throw new BindingConfigParseException(
+                        "Unsupported combination, 'on' and 'off' have to be configured together '" + bindingConfig
```
Use parameterized logging instead of string concatenation.

Markinus replied:
It's a throw string here.

9037568 replied:
You're right. Disregard ;)

9037568 commented on distribution/openhabhome/configurations/openhab_default.cfg:
```diff
@@ -2652,3 +2651,14 @@ tcp:refreshinterval=250
 #
 # Controller model, required, either '0BA7' for LOGO7 or '0BA8' for LOGO8
 #plclogo:bcu.model=0BA8
+###################################### KM200 Binding ##########################################
```
There should be comments indicating which settings are required and which are optional.

If they're all required or they're all optional, a single note near the top stating this would suffice.

9037568 commented on distribution/openhabhome/configurations/openhab_default.cfg:
```diff
@@ -2652,3 +2651,14 @@ tcp:refreshinterval=250
+###################################### KM200 Binding ##########################################
+# The local network ip for the KM200 device
+# km200:ip4_address=192.168.XXX.XXX
+# There a two ways of de-/encription password handling:
```
"encryption"

Markinus commented on bundles/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200Binding.java:
```diff
@@ -0,0 +1,238 @@
+            return;
+        } else {
+            if (config.isEmpty()) {
+                throw new RuntimeException("No properties in openhab.cfg set!");
```
I get this Info an my other pull request of the ddwrt binding. Sometimes is the config null. It seems to be ok.

watou replied:
> is it even possible to have an empty config that is not null?

[This doc](https://osgi.org/javadoc/r4v42/org/osgi/service/cm/ManagedService.html) says the Dictionary can be `null`.  It seems possible that it can be an empty `Dictionary`, but there's no reason I can see to handle that differently than missing required/optional config properties.  When the `Dictionary` passed is `null`, just return from updated with no action or log.

On the subject of config properties, please replace all cases of casting the property value to a String, but instead do `Objects.toString(config.get("X"), null)`, since configuring a binding from the console or via Paper UI can put non-`String` config property values in the `Dictionary`.

Markinus commented:
Ok, I changed all the stuff. Check again, please.

Markinus commented:
I will now start the wiki page.

9037568 commented:
There are a number of change requests you have not yet addressed. Please see the list above.

Also, after reading the documentation, I see that the configuration is required for this binding. That being the case, you should update the properties to make the config actually be required, so that the updated() method will never receive a null config.

Markinus commented on bundles/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200Comm.java:
```diff
@@ -0,0 +1,712 @@
+        String str = "";
+        for (int i = 0; i < len; i++) {
+            if ((data[i] & 0xFF) < 16) {
+                str = str + "0" + java.lang.Integer.toHexString(data[i] & 0xFF);
```
Ok. Changed.

Markinus commented on bundles/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200Comm.java:
```diff
@@ -0,0 +1,712 @@
+                    byte[] test = new byte[1];
+                    return test;
+                } else {
+                    logger.error("HTTP GET failed: " + method.getStatusLine());
```
Ok. Changed.

Markinus commented:
Ok, I will do the changes and the push in.

Markinus commented:
Ok, I changed this things. 
But, I couldn't change the configuration to required. If I do it then the update() method is not more called. (The configuration is set). I can see that the binding is activated but the update method is not called. Any Ideas?

Markinus commented:
Seems to be same Problem like here.. a bug?
https://groups.google.com/forum/#!topic/openhab/kuF5zvMgdfU
https://github.com/openhab/openhab/pull/2028
Most of the bindings did not set this config flag.

9037568 commented:
That's strange. Perhaps @watou can comment on this, since he's the one that raised the issue that you've referenced.

John, it appears that for this binding, if we set the configuration-policy to "require", the binding's start method (KM200Binding.activate()) never gets called... ?

Markinus commented:
Every object is handled in same way, it looks very similar, sometimes it have to be a float and sometimes a string. So it's always the same way. But yes, without would it be possible too.

Markinus commented:
I think I'm done with all request changes. Give me a message if I overlook something.

9037568 commented:
Thanks, @Markinus !

kaikreuzer commented:
This broke the build as it is not Java7 compatible - see https://openhab.ci.cloudbees.com/job/openHAB1-Addons/1379/consoleFull
I'll revert the commit for the moment, please provide a fixed version.

Markinus commented:
hi,

I changed now my default java to 7 and the eclipse settings too. Its 
shouldn't happen again.

I fixed it and started a new pull reqest. It is based on my old branch. 
If you need a complete new commit (all stuff together in one) then give 
me a message, please.

https://github.com/openhab/openhab/pull/4944




Am 28.12.2016 um 17:56 schrieb Kai Kreuzer:
>
> This broke the build as it is not Java7 compatible - see 
> https://openhab.ci.cloudbees.com/job/openHAB1-Addons/1379/consoleFull
> I'll revert the commit for the moment, please provide a fixed version.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/openhab/openhab/pull/4914#issuecomment-269505509>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AXHLy_lRlr30FiSvVFvZW4u2N1UNkeccks5rMpSygaJpZM4LRMxw>.
>
